lib/ansible/modules/packaging/os/redhat_repository.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/python
+
+# Giovanni Sciortino (@giovannisciortino)
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: redhat_repository
+short_description: Manage RHSM repositories using the C(subscription-manager)
+                   command
+description:
+  - Manage(List/Enable/Disable) RHSM repositories to the Red Hat Subscription
+    Management entitlement platform using the C(subscription-manager) command.
+version_added: '2.4'
+author: Giovanni Sciortino (@giovannisciortino)
+notes:
+  - In order to manage RHSM repositories the system must be already registered
+    to RHSM manually or using the ansible module redhat_subscription.
+  - One option between name and list must be defined, both options in the
+    same task must not be defined.
+
+requirements:
+  - subscription-manager
+options:
+  state:
+    description:
+      - Indicates the desired repository state.
+    choices: [enabled, disabled]
+    required: False
```

Comment:
required is usually defined only when true, it defaults to false so there is no need to put it explicitly

Reply:
Not true. The `required` should always be present as per the [docs](http://ansible-docs.readthedocs.io/zh/stable-2.0/rst/developing_modules.html#module-checklist).

Reply:
@jtyr You are right! `required should always be present, be it true or false`

It wasn't the case the last time I read those docs

Reply:
So the documentation contradicts itself, which one is the most recent? [developing_modules_documenting](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html)
`required: Only needed if true, otherwise it is assumed to be false.`

Reply:
Well it depends whether you talk about the `required: False` in the `DOCUMENTATION` block OR about the `required=False` in the `AnsibleModule()` object in the `main()` function. In the first case it must be always present, in the second case only if it's `True`.

Reply:
the `list` option breaks how other modules deal with this a) always return the info b) _facts module c) list/query state (moving away from this).

Reply:
in a) i would return the full list with the 'status' of each repo

Reply:
I hope to have correctly understood the suggestion of @bcoca in his last comment, I have applied the solution a) in the last commit #1a810b2 . 
Now the result variable "repositories"(containing the updated repository states) is returned always from this module when is used with any parameter(name or list).

Reply:
+1 ... for brownie points add --diff support.

Reply:
Added --diff support in the last commit
```
Example:
$ ./bin/ansible gs-client01 -m redhat_repository -a "state=enabled name='rhel-7-server-rpms'" -b --diff 
sudo
-H -S -n
--- before: RHSM repositories
+++ after: RHSM repositories
@@ -1 +1 @@
-Repository 'rhel-7-server-rpms' is disabled for this system
+Repository 'rhel-7-server-rpms' is enabled for this system

gs-client01 | SUCCESS => {
    "changed": true, 
    "failed": false, 
    "repositories": [
        {
            "enabled": false, 
            "id": "rhel-7-server-optional-rpms", 
            "name": "Red Hat Enterprise Linux 7 Server - Optional (RPMs)", 
            "url": "https://gs-sat61.localdomain/pulp/repos/Org1/Library/content/dist/rhel/server/7/$releasever/$basearch/optional/os"
        }, 
        {
            "enabled": false, 
            "id": "rhel-7-server-satellite-tools-6.2-rpms", 
            "name": "Red Hat Satellite Tools 6.2 (for RHEL 7 Server) (RPMs)", 
            "url": "https://gs-sat61.localdomain/pulp/repos/Org1/Library/content/dist/rhel/server/7/7Server/$basearch/sat-tools/6.2/os"
        }, 
        {
            "enabled": true, 
            "id": "rhel-7-server-rpms", 
            "name": "Red Hat Enterprise Linux 7 Server (RPMs)", 
            "url": "https://gs-sat61.localdomain/pulp/repos/Org1/Library/content/dist/rhel/server/7/$releasever/$basearch/os"
        }
    ], 
    "results": [
        "Repository 'rhel-7-server-rpms' is enabled for this system."
    ]
}
```

Removed list option, now the option state accepts the following parameters: enabled, disabled, list, list_enabled, list_disabled. When state is equal to enabled or disabled, the option name allow to define the list of repositories to be enable/disable.